src/Mono.Android/Android.Runtime/JNIEnv.cs

```diff
@@ -632,13 +638,49 @@ public static string GetClassNameFromInstance (IntPtr jobject)
 		}
 
 		[DllImport ("__Internal", CallingConvention = CallingConvention.Cdecl)]
-		internal static extern IntPtr monodroid_typemap_managed_to_java (string managed);
+		static extern IntPtr monodroid_typemap_managed_to_java (byte[] mvid, int token);
+
+		internal static void LogTypemapTrace (StackTrace st)
+		{
+			string trace = st.ToString ()?.Trim ();
+			if (String.IsNullOrEmpty (trace))
+				return;
+
+			monodroid_log (LogLevel.Warn, LogCategories.Assembly, "typemap: called from");
+			foreach (string line in trace.Split ('\n')) {
+				monodroid_log (LogLevel.Warn, LogCategories.Assembly, line);
+			}
+		}
+
+		internal static IntPtr TypemapManagedToJava (Type type)
+		{
+			if (mvid_bytes == null)
+				mvid_bytes = new byte[16];
+
+			Span<byte> mvid = new Span<byte>(mvid_bytes);
+			byte[] mvid_slow = null;
+			if (!type.Module.ModuleVersionId.TryWriteBytes (mvid)) {
+				monodroid_log (LogLevel.Warn, LogCategories.Default, $"Failed to obtain module MVID using the fast method, falling back to the slow one");
+				mvid_slow = type.Module.ModuleVersionId.ToByteArray ();
+			}
+
+			IntPtr ret = monodroid_typemap_managed_to_java (mvid_slow == null ? mvid_bytes : mvid_slow, type.MetadataToken);
+
+			if (ret == IntPtr.Zero) {
+				monodroid_log (LogLevel.Warn, LogCategories.Default, $"typemap: failed to map managed type to Java type: {type.AssemblyQualifiedName} (Module ID: {type.Module.ModuleVersionId}; Type token: {type.MetadataToken})");
```

Comment:
The question is, should this message be printed when `LogTypemapMissingStackTrace` is false?  Is there any reason why we'd want this message to ~always be printed, particularly when during the development of this PR we've run across various "corner cases" in which we'd try to "typemap" `System.Int32`, which will *never* work, but is "okay" (-ish) and doesn't *break* anything or otherwise indicate that a "problem" has occurred.

Rephrased: should we be spamming logcat with this message unless it's explicitly requested in the first place?